### PR TITLE
Enable cusolver potrf batched for Cholesky decomposition when cuda >= 11.3

### PR DESCRIPTION
**Pull request by xwang233**
This PR enables the usage of cusolver potrf batched as the backend of Cholesky decomposition (`torch.linalg.cholesky` and `torch.linalg.cholesky_ex`) when cuda version is greater than or equal to 11.3.

Benchmark available at https://github.com/xwang233/code-snippet/tree/master/linalg/cholesky-new. It is seen that cusolver potrf batched performs better than magma potrf batched in most cases.

### before:

- batch size == 1: cusolver potrf
- batch size > 1: magma xpotrf batched

### after:

cuda >= 11.3:
- batch size == 1: cusolver potrf
- batch size > 1: cusolver potrf batched

cuda < 11.3 (not changed):
- batch size == 1: cusolver potrf
- batch size > 1: magma xpotrf batched

--- 

See also #42666 #47953 #53104 #53879

### CODE REVIEW

**Comment by xwang233** [comment]
reserved

**Comment by IvanYashchuk** [comment]
If both cuSOLVER and MAGMA are available and CUDA version is < 11.3 we should continue using batched MAGMA as it has better performance, than single input cuSOLVER variant called in a loop, right? This PR modifies the behavior for < 11.3 versions to use looped cuSOLVER instead of batched MAGMA.

Besides that dispatch issue, everything looks good.

**Comment by xwang233** [comment]
Ohhh, yes, you're right. Let me fix that dispatch logic. 😄

File: aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu [hunk 1]
```diff
@@ -562,10 +559,41 @@ inline static void apply_cholesky_cusolver(const Tensor& self_working_copy, bool
 #endif // USE_CUSOLVER_64_BIT
 }
 
+// Implementation of Cholesky decomposition using batched cusolverDn<T>potrfBatched
+// Warning: cusolverDn<T>potrfBatched doesn't work quite well when matrix size or batch size is zero.
+// If you write your own C++ extension and use this function, make sure you do a zero numel check for the input.
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraLib.cu:564]: Nice note

File: aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h [hunk 2]
```diff
@@ -12,6 +12,15 @@
 #define USE_CUSOLVER
 #endif
 
+// cusolverDn<T>potrfBatched may have numerical issue before cuda 11.3 release,
```
Comment (mruberry) [inline aten/src/ATen/native/cuda/BatchLinearAlgebraLib.h:15]: Great comment

**Review by mruberry — APPROVED** [review]
An excellent CUDA performance PR to cap the many performance improvements realized for the release of torch.linalg. in PyTorch 1.9

cc @ptrblck

**Pull request closed without merging**